### PR TITLE
Remove support for Databricks-9.1 ML LTS [databricks]

### PR DESCRIPTION
**Pull request by nartal1**
This closes https://github.com/NVIDIA/spark-rapids/issues/7463

This removes support for databricks-9.1 ML LTS (312db shim).
Using the shimplify (https://github.com/NVIDIA/spark-rapids/blob/branch-23.04/docs/dev/shimplify.md#deleting-a-shim) method to delete the shim which removes all files under spark312db directory and rearranges the remaining files in appropriate directories.

Updated pom, integration tests, build scripts and documentation which  referenced 312db/Databricks-9.1.

### CODE REVIEW

**Comment by nartal1** [comment]
build

File: integration_tests/src/main/python/arithmetic_ops_test.py [hunk 1]
```diff
@@ -1039,7 +1039,7 @@ def test_div_overflow_exception_when_ansi(expr, ansi_enabled):
 
 # Only run this test before Spark v3.2.0 to verify IntegralDivide will NOT
 # throw exceptions for overflow even ANSI mode is enabled.
-@pytest.mark.skipif(not is_before_spark_320() or is_databricks91_or_later(), reason='https://github.com/apache/spark/pull/32260')
+@pytest.mark.skipif(not is_before_spark_320() or is_databricks104_or_later(), reason='https://github.com/apache/spark/pull/32260')
```
Comment (jlowe) [inline integration_tests/src/main/python/arithmetic_ops_test.py:1042]:
```suggestion
@pytest.mark.skipif(not is_before_spark_320(), reason='https://github.com/apache/spark/pull/32260')
```

File: integration_tests/src/main/python/cast_test.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
```
Comment (jlowe) [inline integration_tests/src/main/python/cast_test.py:16]: 2023 copyrights

File: integration_tests/src/main/python/cast_test.py [hunk 3]
```diff
@@ -88,7 +88,7 @@ def test_cast_string_date_valid_format():
 # Spark 320+ and databricks support Ansi mode when casting string to date
 # This means an exception will be thrown when casting invalid string to date on Spark 320+ or databricks
 # test Spark versions < 3.2.0 and non databricks, ANSI mode
-@pytest.mark.skipif((not is_before_spark_320()) or is_databricks91_or_later(), reason="ansi cast(string as date) throws exception only in 3.2.0+ or db")
+@pytest.mark.skipif((not is_before_spark_320()) or is_databricks104_or_later(), reason="ansi cast(string as date) throws exception only in 3.2.0+ or db")
```
Comment (jlowe) [inline integration_tests/src/main/python/cast_test.py:91]:
```suggestion
@pytest.mark.skipif(not is_before_spark_320(), reason="ansi cast(string as date) throws exception only in 3.2.0+ or db")
```

File: integration_tests/src/main/python/cast_test.py [hunk 4]
```diff
@@ -97,7 +97,7 @@ def test_cast_string_date_invalid_ansi_before_320():
               'spark.sql.ansi.enabled': 'true'}, )
 
 # test databricks, ANSI mode, all databricks versions supports Ansi mode when casting string to date
-@pytest.mark.skipif(not is_databricks91_or_later(), reason="Spark versions(< 320) not support Ansi mode when casting string to date")
+@pytest.mark.skipif(not is_databricks104_or_later(), reason="Spark versions(< 320) not support Ansi mode when casting string to date")
 @pytest.mark.parametrize('invalid', invalid_values_string_to_date)
 def test_cast_string_date_invalid_ansi_databricks(invalid):
```
Comment (jlowe) [inline integration_tests/src/main/python/cast_test.py:102]: Seems like this test is now redundant, should be covered by test_cast_string_date_invalid_ansi below.

File: integration_tests/src/main/python/cast_test.py [hunk 5]
```diff
@@ -107,7 +107,7 @@ def test_cast_string_date_invalid_ansi_databricks(invalid):
         error_message="DateTimeException")
 
 # test databricks, ANSI mode, valid values
-@pytest.mark.skipif(not is_databricks91_or_later(), reason="Spark versions(< 320) not support Ansi mode when casting string to date")
+@pytest.mark.skipif(not is_databricks104_or_later(), reason="Spark versions(< 320) not support Ansi mode when casting string to date")
```
Comment (jlowe) [inline integration_tests/src/main/python/cast_test.py:110]: Shouldn't this test apply to everything >= Spark 3.2 and not just Databricks?  I didn't see an equivalent test for non-Databricks with Spark 3.2+, seems to be a dropped test case for those platforms.

File: integration_tests/src/main/python/delta_lake_test.py [hunk 6]
```diff
@@ -16,15 +16,13 @@
 from pyspark.sql import Row
```
Comment (jlowe) [inline integration_tests/src/main/python/delta_lake_test.py:16]: 2023 copyrights

File: integration_tests/src/main/python/spark_session.py [hunk 7]
```diff
@@ -184,9 +184,6 @@ def is_databricks_version_or_later(major, minor):
     db_minor = int(parts[1])
```
Comment (jlowe) [inline integration_tests/src/main/python/spark_session.py:184]: 2023 copyrights

File: jenkins/databricks/deploy.sh [hunk 8]
```diff
@@ -25,7 +25,7 @@ echo "Maven mirror is $MVN_URM_MIRROR"
 SERVER_ID='snapshots'
```
Comment (jlowe) [inline jenkins/databricks/deploy.sh:25]: 2023 copyrights

File: jenkins/databricks/params.py [hunk 9]
```diff
@@ -23,7 +23,7 @@
 script_dest = '/home/ubuntu/build.sh'
```
Comment (jlowe) [inline jenkins/databricks/params.py:23]: 2023 copyrights

**Review by jlowe — COMMENTED** [review]
Nit: the places using is_databricks104_or_later should be using is_datarbicks_runtime instead.

File: integration_tests/src/main/python/cast_test.py (hunk 4 above)
Comment (nartal1) [reply integration_tests/src/main/python/cast_test.py]: removed this test

File: integration_tests/src/main/python/cast_test.py (hunk 5 above)
Comment (nartal1) [reply integration_tests/src/main/python/cast_test.py]: You are right. Updated the test to run for Spark-3.2+ . PTAL.

**Comment by nartal1** [comment]
build

File: integration_tests/src/main/python/cast_test.py [hunk 10]
```diff
@@ -16,7 +16,7 @@
 
 from asserts import assert_gpu_and_cpu_are_equal_collect, assert_gpu_and_cpu_are_equal_sql, assert_gpu_and_cpu_error, assert_gpu_fallback_collect, assert_py4j_exception
 from data_gen import *
-from spark_session import is_before_spark_320, is_before_spark_330, is_databricks91_or_later, with_gpu_session
+from spark_session import is_before_spark_320, is_before_spark_330, is_databricks104_or_later, with_gpu_session
```
Comment (jlowe) [inline integration_tests/src/main/python/cast_test.py:19]:
```suggestion
from spark_session import is_before_spark_320, is_before_spark_330, with_gpu_session
```

File: integration_tests/src/main/python/delta_lake_test.py [hunk 11]
```diff
@@ -16,15 +16,13 @@
...
-from spark_session import with_cpu_session, with_gpu_session, is_databricks91_or_later, is_databricks104_or_later, \
-    spark_version
+from spark_session import with_cpu_session, with_gpu_session, is_databricks104_or_later, spark_version
 
 _conf = {'spark.rapids.sql.explain': 'ALL'}
 
 @delta_lake
 @allow_non_gpu('FileSourceScanExec')
-@pytest.mark.skipif(not (is_databricks91_or_later() or spark_version().startswith("3.2.")), \
+@pytest.mark.skipif(not (is_databricks104_or_later() or spark_version().startswith("3.2.")), \
```
Comment (jlowe) [inline integration_tests/src/main/python/delta_lake_test.py:25]: Nit: This, and elsewhere, should be using is_databricks_runtime() instead.

**Comment by nartal1** [comment]
build

**Review by jlowe — APPROVED** [review]

**Pull request merged**